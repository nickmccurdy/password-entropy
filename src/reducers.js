import GenericOptions from './options/GenericOptions'
import { assoc, identity, map, T } from 'ramda'
import { combineReducers } from 'redux'

function createReducer(actionType, initialState, callback = identity) {
  return (state = initialState, { payload, type }) => {
    return type === actionType ? callback(payload, state) : state
  }
}

export default combineReducers({
  length: createReducer('SET_LENGTH', 6),
  options: combineReducers({
    custom: createReducer('SET_CUSTOM_POSSIBLE_ITEMS', 1),
    generic: createReducer ('TOGGLE_GENERIC',
                            map(T, GenericOptions.toggles),
                            ({ checked, name }, state) => assoc(name, checked, state))
  }),
  optionsIndex: createReducer('SET_OPTIONS_INDEX', 0)
})
