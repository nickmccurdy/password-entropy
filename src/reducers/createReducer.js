import { identity } from 'ramda'

export default function createReducer(actionType, initialState, callback = identity) {
  return (state = initialState, { payload, type }) => {
    const matchesAction = type === actionType
    const isValid = typeof payload !== 'number' || payload >= 0

    return matchesAction && isValid ? callback(payload, state) : state
  }
}
