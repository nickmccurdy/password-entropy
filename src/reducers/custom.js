import { path } from 'ramda'

import createReducer from './createReducer'

export default createReducer('SET_CUSTOM', 0)

export function setCustom({ target: { value } }) {
  return { type: 'SET_CUSTOM', payload: parseInt(value, 10) }
}

export const selector = state => ({
  possibleItems: path(['options', 'custom'], state)
})
