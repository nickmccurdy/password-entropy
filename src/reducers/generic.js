import assoc from 'ramda/src/assoc'
import map from 'ramda/src/map'
import path from 'ramda/src/path'
import T from 'ramda/src/T'

import createReducer from './createReducer'

export const toggles = {
  Letters: {
    example: 'a-z',
    possibleItems: 26
  },
  'Capital Letters': {
    example: 'A-Z',
    possibleItems: 26
  },
  Numbers: {
    example: '0-9',
    possibleItems: 10
  },
  Symbols: {
    example: '!@#$%^&*',
    possibleItems: 8
  }
}

export const defaultToggles = map(T, toggles)

export function toggleGenericCallback({ checked, name }, state) {
  return assoc(name, checked, state)
}

export default createReducer(
  'TOGGLE_GENERIC',
  defaultToggles,
  toggleGenericCallback
)

export function toggleGeneric({ target: { name } }, checked) {
  return {
    type: 'TOGGLE_GENERIC',
    payload: {
      checked,
      name
    }
  }
}

export const selector = path(['options', 'Generic'])
