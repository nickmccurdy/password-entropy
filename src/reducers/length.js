import pick from 'ramda/src/pick'

import createReducer from './createReducer'

export default createReducer('SET_LENGTH', 12)

export function setLength({ target: { value } }) {
  return {
    type: 'SET_LENGTH',
    payload: parseInt(value, 10)
  }
}

export const selector = pick(['length'])
