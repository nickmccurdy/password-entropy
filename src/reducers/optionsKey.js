import pick from 'ramda/src/pick'

import createReducer from './createReducer'

export default createReducer('SET_OPTIONS_KEY', 'Generic')

export function setOptionsKey(event, key) {
  return {
    type: 'SET_OPTIONS_KEY',
    payload: key
  }
}

export const selector = pick(['optionsKey'])
