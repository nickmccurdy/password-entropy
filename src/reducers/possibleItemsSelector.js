import blue from 'material-ui/colors/blue'
import green from 'material-ui/colors/green'
import red from 'material-ui/colors/red'
import yellow from 'material-ui/colors/yellow'
import always from 'ramda/src/always'
import filter from 'ramda/src/filter'
import findLast from 'ramda/src/findLast'
import identity from 'ramda/src/identity'
import keys from 'ramda/src/keys'
import map from 'ramda/src/map'
import pipe from 'ramda/src/pipe'
import sum from 'ramda/src/sum'

import { toggles } from './generic'

export const dicewareWords = 7776

const entropyTips = [
  {
    color: red,
    minimum: 0,
    strength: 'Very Weak'
  },
  {
    color: yellow,
    minimum: 32,
    strength: 'Weak'
  },
  {
    color: blue,
    minimum: 64,
    strength: 'Strong'
  },
  {
    color: green,
    minimum: 128,
    strength: 'Very Strong'
  }
]

const possibleItems = {
  generic: pipe(
    filter(identity),
    keys,
    map(key => toggles[key].possibleItems),
    sum
  ),
  diceware: always(dicewareWords),
  custom: identity
}

export default function possibleItemsSelector({ length, options, optionsKey }) {
  const possiblePasswords = possibleItems[optionsKey](options[optionsKey]) ** length
  const entropyBits = Math.max(0, Math.log2(possiblePasswords))

  return {
    approximate: possiblePasswords > Number.MAX_SAFE_INTEGER,
    entropyBits,
    entropyTip: findLast(tip => entropyBits >= tip.minimum, entropyTips),
    possiblePasswords
  }
}
