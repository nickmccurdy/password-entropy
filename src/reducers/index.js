import { combineReducers } from 'redux'

import custom from './custom'
import generic from './generic'
import length from './length'
import optionsKey from './optionsKey'

export default combineReducers({
  length,
  options: combineReducers({ custom, generic }),
  optionsKey
})
