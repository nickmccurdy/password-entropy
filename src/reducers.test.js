import reducers from './reducers'

describe('reducers', () => {
  it('return initial state', () => {
    expect(reducers(undefined, {})).toMatchSnapshot()
  })

  function dispatch (type, payload) {
    return reducers(undefined, { type, payload })
  }

  it('handle SET_CUSTOM_POSSIBLE_ITEMS', () => {
    expect(dispatch('SET_CUSTOM_POSSIBLE_ITEMS', 2)).toMatchSnapshot()
  })

  it('handle SET_LENGTH', () => {
    expect(dispatch('SET_LENGTH', 7)).toMatchSnapshot()
  })

  it('handle SET_OPTIONS_INDEX', () => {
    expect(dispatch('SET_OPTIONS_INDEX', 1)).toMatchSnapshot()
  })

  it('handle TOGGLE_GENERIC', () => {
    expect(dispatch('TOGGLE_GENERIC', { name: 'letters', checked: false })).toMatchSnapshot()
  })
})
