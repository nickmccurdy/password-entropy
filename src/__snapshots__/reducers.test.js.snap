// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers handle SET_CUSTOM_POSSIBLE_ITEMS 1`] = `
Object {
  "length": 6,
  "options": Object {
    "custom": 2,
    "generic": Object {
      "capitalLetters": true,
      "letters": true,
      "numbers": true,
      "symbols": true,
    },
  },
  "optionsIndex": 0,
}
`;

exports[`reducers handle SET_LENGTH 1`] = `
Object {
  "length": 7,
  "options": Object {
    "custom": 1,
    "generic": Object {
      "capitalLetters": true,
      "letters": true,
      "numbers": true,
      "symbols": true,
    },
  },
  "optionsIndex": 0,
}
`;

exports[`reducers handle SET_OPTIONS_INDEX 1`] = `
Object {
  "length": 6,
  "options": Object {
    "custom": 1,
    "generic": Object {
      "capitalLetters": true,
      "letters": true,
      "numbers": true,
      "symbols": true,
    },
  },
  "optionsIndex": 1,
}
`;

exports[`reducers handle TOGGLE_GENERIC 1`] = `
Object {
  "length": 6,
  "options": Object {
    "custom": 1,
    "generic": Object {
      "capitalLetters": true,
      "letters": false,
      "numbers": true,
      "symbols": true,
    },
  },
  "optionsIndex": 0,
}
`;

exports[`reducers return initial state 1`] = `
Object {
  "length": 6,
  "options": Object {
    "custom": 1,
    "generic": Object {
      "capitalLetters": true,
      "letters": true,
      "numbers": true,
      "symbols": true,
    },
  },
  "optionsIndex": 0,
}
`;
